supernode/daemon/mgr/pieceerror/manager.go

```diff
@@ -0,0 +1,162 @@
+}
+
+func NewManager(cfg *config.Config, gcManager mgr.GCMgr, cdnManager mgr.CDNMgr) (*Manager, error) {
+	return &Manager{
+		cfg:                cfg,
+		gcManager:          gcManager,
+		cdnManager:         cdnManager,
+		pieceErrChan:       make(chan *types.PieceErrorRequest, ErrHandlerChanSize),
+		errorHandlingStore: syncmap.NewSyncMap(),
+		handledStore:       syncmap.NewSyncMap(),
+	}, nil
+}
+
+// HandlePieceError the peer should report the error with related info when
+// it failed to download a piece from supernode.
+// And the supernode should handle the piece Error and do some repair operations.
+func (em *Manager) HandlePieceError(ctx context.Context, pieceErrorRequest *types.PieceErrorRequest) error {
+	// ignore the error that isn't caused by downloading from supernode
+	if em.cfg.IsSuperPID(pieceErrorRequest.DstPid) {
+		return nil
+	}
+
+	// if the error is handling, we should ignore it.
+	if _, err := em.errorHandlingStore.Get(pieceErrorRequest.TaskID); err == nil {
+		if !errortypes.IsDataNotFound(err) {
+			logrus.Warnf("failed to get taskID(%s) from errorHandlingStore: %v", pieceErrorRequest.TaskID, err)
+		}
+		return nil
+	}
+
+	em.errorHandlingStore.Add(pieceErrorRequest.TaskID, true)
+	em.pieceErrChan <- pieceErrorRequest
+
+	return nil
+}
+
+// StartHandleError starts a goroutine to handle the piece error.
+func (em *Manager) StartHandleError(ctx context.Context) {
+	go func() {
+		em.startHandleErrorPool(ctx)
+	}()
+
+	go func() {
+		ticker := time.NewTicker(GCHandlingInterval)
+		for range ticker.C {
+			em.deleteHandling(ctx)
+		}
+	}()
+}
+
+func (em *Manager) deleteHandling(ctx context.Context) {
+	rangeFunc := func(key, value interface{}) bool {
+		if handledTime, ok := value.(time.Time); ok {
+			if time.Since(handledTime) > GCHandlingDelay {
+				if taskID, ok := key.(string); ok {
+					em.errorHandlingStore.Delete(taskID)
+					em.handledStore.Delete(taskID)
+				}
+			}
+		}
+		return true
+	}
+
+	em.handledStore.Range(rangeFunc)
+}
+func (em *Manager) startHandleErrorPool(ctx context.Context) {
+	var wg = &sync.WaitGroup{}
+	for i := 0; i < HandleErrorPool; i++ {
+		wg.Add(1)
+		go func() {
+			for per := range em.pieceErrChan {
+				if err := em.handleError(ctx, per); err != nil {
+					logrus.Errorf("failed to handle error %+v:%v", per, err)
+				}
+			}
+			wg.Done()
+		}()
+	}
+}
+
+func (em *Manager) handleError(ctx context.Context, pieceError *types.PieceErrorRequest) error {
+	handler, err := em.getHandler(ctx, pieceError.ErrorType)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get handler")
+	}
+
+	if err := handler.Handle(ctx, pieceError); err != nil {
+		return err
+	}
+
+	em.handledStore.Add(pieceError.TaskID, time.Now())
+	return nil
+}
+
+func (em *Manager) getHandler(ctx context.Context, errType string) (Handler, error) {
+	switch errType {
+	case types.PieceErrorRequestErrorTypeFILEMD5NOTMATCH:
+		return newFileMd5NotMatchHandler(em.gcManager, em.cdnManager), nil
```

Comment:
It's better to use a created object before rather than creating a new object every time.